File: ordenacao-busca/python/busca_binaria_recursiva.py
from typing import List

#   Problema: Busca Binária (Recursiva)
#
#   n = number of elements array
#
#   time: O(log n)
#   space: O(1)
#

def busca_binaria(nums: List[int], key: int, low: int, high: int) -> int:
    if low > high:
        return -1
    
    middle = (low + high) // 2
    if key == nums[middle]:
        return middle
    elif key < nums[middle]:
        return busca_binaria(nums, key, low, middle - 1)
    else:
        return busca_binaria(nums, key, middle + 1, high)


nums = [-5, 0, 2, 8, 13, 16, 19, 23, 29, 34, 38]
key = 34

result = busca_binaria(nums, key, 0, len(nums)) # 9

if result != -1:
    print(f"Elemento {key} encontrado na posicao {result}.")
else:
    print(f"Elemento {key} não encontrado no array.")

nums = [-5, 0, 2, 8, 13, 16, 19, 23, 29, 34, 38]
key = 10

result = busca_binaria(nums, key, 0, len(nums)) # -1

if result != -1:
    print(f"Elemento {key} encontrado na posicao {result}.")
else:
    print(f"Elemento {key} não encontrado no array.")


nums = [-10, -3, 4, 11, 13, 18, 44, 64, 91, 225, 431]
key = 11

result = busca_binaria(nums, key, 0, len(nums)) # 3

if result != -1:
    print(f"Elemento {key} encontrado na posicao {result}.")
else:
    print(f"Elemento {key} não encontrado no array.")

File: ordenacao-busca/csharp/busca_binaria_recursiva.cs
using System;

/*
 *  Problema: Busca binária (Iterativa)
 * 
 *  n = number of elements array
 * 
 *  time: O(log n)
 *  space: O(n)
 */

public class busca_binaria_recursiva {

    public static int BuscaBinaria(int[] nums, int key, int low, int high) {
        if (low > high) {
            return -1;
        }

        int middle = (low + high) / 2;
        if (key == nums[middle]) {
            return middle;
        }
        else if (key < nums[middle]) {
            return BuscaBinaria(nums, key, low, middle - 1);
        }
        else {
            return BuscaBinaria(nums, key, middle + 1, high);
        }
    }

    static void Main() {
        int[] nums = {-5, 0, 2, 8, 13, 16, 19, 23, 29, 34, 38};
        int key = 34;
        Console.WriteLine(BuscaBinaria(nums, key, 0, nums.Length));
        
        int[] nums1 = {-5, 0, 2, 8, 13, 16, 19, 23, 29, 34, 38};
        int key1 = 10;
        Console.WriteLine(BuscaBinaria(nums1, key1, 0, nums1.Length));
    }
    
}